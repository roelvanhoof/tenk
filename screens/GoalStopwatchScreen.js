import * as React from 'react'
import { View, Text, StyleSheet, Alert, AsyncStorage } from 'react-native'
import { Stopwatch } from 'react-native-stopwatch-timer'
import { Button } from 'react-native-elements'

const options = {
  container: {
    marginBottom: 34,
    alignSelf: 'center',
  },
  text: {
    fontSize: 74,
  },
}

const styles = StyleSheet.create({
  container: {
    paddingTop: 104,
    flex: 1,
    alignItems: 'stretch',
    backgroundColor: '#fff',
  },
  button: {
    width: 140,
    marginBottom: 16,
    alignSelf: 'center',
  },
  text: {
    marginTop: 100,
    width: 340,
    alignSelf: 'center',
    textAlign: 'center',
  },
  text2: {
    marginTop: 16,
    width: 340,
    alignSelf: 'center',
    textAlign: 'center',
  },
  text3: {
    marginTop: 16,
    width: 340,
    alignSelf: 'center',
    textAlign: 'center',
  },
})

export default function GoalStopwatchScreen({ route, navigation }) {
  const { goal } = route.params

  const [stopwatchRunning, setStopwatchRunning] = React.useState(false)
  const [stopwatchReset, setStopwatchReset] = React.useState(false)
  const [formattedTime, setFormattedTime] = React.useState('00:00:00')

  React.useEffect(() => {
    navigation.setOptions({
      headerTitle: goal.name,
      headerBackTitle: 'Back',
    })
  }, [goal])

  function timeToSeconds() {
    const a = formattedTime.split(':')
    return +a[0] * 60 * 60 + +a[1] * 60 + +a[2]
  }

  async function addTimeToGoal() {
    const seconds = timeToSeconds()
    const goalsString = await AsyncStorage.getItem('@goals')
    const goals = JSON.parse(goalsString)
    let goalFound = false
    for (let i = 0; i < goals.length; i += 1) {
      if (goals[i].id === goal.id) {
        goals[i].spent += seconds
        goalFound = true
      }
    }
    if (!goalFound) {
      throw new Error('Goal not found')
    }
    await AsyncStorage.setItem('@goals', JSON.stringify(goals))
    navigation.goBack()
  }

  function toggleStopwatch() {
    if (stopwatchRunning) {
      setStopwatchRunning(false)
      Alert.alert(
        'Update Goal?',
        `Do you want to add ${formattedTime} towards goal "${goal.name}"`,
        [
          {
            text: 'Cancel',
            style: 'cancel',
          },
          {
            text: 'OK',
            onPress: () => {
              setStopwatchReset(true)
              return addTimeToGoal()
            },
          },
        ]
      )
    } else {
      setStopwatchRunning(true)
    }
  }

  function resetStopwatch() {
    setStopwatchRunning(false)
    setStopwatchReset(true)
  }

  return (
    <View style={styles.container}>
      <Stopwatch
        laps
        start={stopwatchRunning}
        reset={stopwatchReset}
        options={options}
        getTime={time => setFormattedTime(time)}
      />
      <Button
        title={!stopwatchRunning ? 'Start' : 'Stop'}
        onPress={toggleStopwatch}
        style={styles.button}
      />
      <Button title="Reset" onPress={resetStopwatch} style={styles.button} />
      <Text style={styles.text}>
        Start the stopwatch and start working on you goal.
      </Text>
      <Text style={styles.text2}>
        When you are finished the time will be added towards completing your
        goal.
      </Text>
      <Text style={styles.text3}>Good luck, you can do it!</Text>
    </View>
  )
}
