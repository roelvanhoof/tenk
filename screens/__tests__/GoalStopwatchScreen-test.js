import * as React from 'react'
import renderer from 'react-test-renderer'
import { render, act, fireEvent } from 'react-native-testing-library'

import { AsyncStorage } from 'react-native'

import { NavigationContainer } from '@react-navigation/native'
import { createStackNavigator } from '@react-navigation/stack'

import GoalStopwatchScreen from '../GoalStopwatchScreen'

const Stack = createStackNavigator()

jest.mock('../../navigation/getGoalId')

jest.mock('react-native/Libraries/Storage/AsyncStorage', () => ({
  setItem: jest.fn(() => {
    return new Promise(resolve => {
      resolve(null)
    })
  }),
  getItem: jest.fn(() => {
    return new Promise(resolve => {
      const goals = [
        {
          id: '123',
          name: 'Learn French',
          hours: 10000,
          spent: 4000 * 3600,
        },
      ]
      resolve(JSON.stringify(goals))
    })
  }),
}))

beforeEach(() => {
  jest.clearAllMocks()
})

describe('GoalStopwatchScreen', () => {
  it(`renders correctly`, () => {
    // const container = (
    //   <NavigationContainer>
    //     <Stack.Navigator>
    //       <Stack.Screen name="GoalStopwatch" component={GoalStopwatchScreen} />
    //     </Stack.Navigator>
    //   </NavigationContainer>
    // )
    const navigation = {
      navigate: jest.fn(),
      setOptions: jest.fn(),
    }
    const route = {
      params: {
        goal: {
          id: '123',
          name: 'Learn french',
        },
      },
    }
    const tree = renderer
      .create(<GoalStopwatchScreen navigation={navigation} route={route} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
