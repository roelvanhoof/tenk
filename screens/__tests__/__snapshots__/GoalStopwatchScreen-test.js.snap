// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalStopwatchScreen renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingTop": 104,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "marginBottom": 34,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 74,
        }
      }
    >
      00:00:00
    </Text>
  </View>
  <View
    style={
      Object {
        "borderRadius": 3,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "marginBottom": 16,
          "opacity": 1,
          "width": 140,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          Start
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderRadius": 3,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "marginBottom": 16,
          "opacity": 1,
          "width": 140,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          Reset
        </Text>
      </View>
    </View>
  </View>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 100,
        "textAlign": "center",
        "width": 340,
      }
    }
  >
    Start the stopwatch and start working on you goal.
  </Text>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 16,
        "textAlign": "center",
        "width": 340,
      }
    }
  >
    When you are finished the time will be added towards completing your goal.
  </Text>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 16,
        "textAlign": "center",
        "width": 340,
      }
    }
  >
    Good luck, you can do it!
  </Text>
</View>
`;
